crates/lowering/src/ast_utils.rs

```diff
@@ -0,0 +1,344 @@
+use fe_parser::ast::{CallArg, Expr, FuncStmt};
+use fe_parser::node::Node;
+
+use crate::utils::ZeroSpanNode;
+
+pub enum StmtOrExpr {
+    Stmt(Node<FuncStmt>),
+    Expr(Node<Expr>),
+}
+
+impl From<Node<FuncStmt>> for StmtOrExpr {
+    fn from(stmt: Node<FuncStmt>) -> Self {
+        StmtOrExpr::Stmt(stmt)
+    }
+}
+
+impl From<Node<Expr>> for StmtOrExpr {
+    fn from(expr: Node<Expr>) -> Self {
+        StmtOrExpr::Expr(expr)
+    }
+}
+
+impl StmtOrExpr {
+    pub fn as_stmt(&self) -> Node<FuncStmt> {
+        match self {
+            StmtOrExpr::Stmt(stmt) => stmt.clone(),
+            _ => panic!("meh"),
+        }
+    }
+
+    pub fn as_expr(&self) -> Node<Expr> {
+        match self {
+            StmtOrExpr::Expr(expr) => expr.clone(),
+            _ => panic!("meh"),
+        }
+    }
+}
+
+pub fn map_ast_node<S>(thingy: StmtOrExpr, map_fn: &mut S) -> StmtOrExpr
+where
+    S: FnMut(StmtOrExpr) -> StmtOrExpr,
+{
+    // TODO: This needs to walk the entire tree. Currently only walks enough to get my tests passing
```

Comment:
@sbillig warning, the code is very WIP and it doesn't yet walk the entire tree (only enough to get my tests passing).

So, the function that you asked about was `fn function_static_strings(db, function: FunctionId) -> Rc<IndexSet<String>>`. I think that you *could* build this on top of this if you call `map_ast_node` for each function statement and if the passed closure accumulates the `IndexSet<String>` as a side effect. That said, it's not particular elegant to have a mapping function leak out a side effect into a variable that's wrapped into the closure...I think it would actually be nicer to have another function `scan` that can traverse the tree and accumulate information into a single return value..but I haven't gotten that far yet :grimacing: